Implement URL.parse()

annevk opened the pull request:
#### e1614bd8c992e7e6e162fff5143f3a4932226ae5
<pre>
Implement URL.parse()
<a href="https://bugs.webkit.org/show_bug.cgi?id=271636">https://bugs.webkit.org/show_bug.cgi?id=271636</a>

Reviewed by Alex Christensen.

This was standardized in <a href="https://github.com/whatwg/url/pull/825">https://github.com/whatwg/url/pull/825</a> and test
coverage was added here:
<a href="https://github.com/web-platform-tests/wpt/pull/45248">https://github.com/web-platform-tests/wpt/pull/45248</a>

As a drive-by fix we remove m_baseURL from DOMURL as it does not need it.

* LayoutTests/imported/w3c/web-platform-tests/url/url-statics-parse.any-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/url/url-statics-parse.any.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/url/url-statics-parse.any.js: Added.
(forEach):
(test):
* LayoutTests/imported/w3c/web-platform-tests/url/url-statics-parse.any.worker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/url/url-statics-parse.any.worker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/url/w3c-import.log:
* Source/WebCore/html/DOMURL.cpp:
(WebCore::DOMURL::DOMURL):
(WebCore::DOMURL::create):
(WebCore::parseInternal):
(WebCore::DOMURL::parse):
(WebCore::DOMURL::canParse):
* Source/WebCore/html/DOMURL.h:
* Source/WebCore/html/DOMURL.idl:

Canonical link: <a href="https://commits.webkit.org/276656@main">https://commits.webkit.org/276656@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4d52cd678b3570e1310b3958c90bfaa54d946525

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/45247 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/24365 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/47767 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/47910 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/41255 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/47554 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/28554 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/21762 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/37113 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/45825 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/21414 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/39025 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/18211 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/18827 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/40107 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/3294 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/41510 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/40424 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/49615 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/20229 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/16760 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/44146 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/21535 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/39191 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/42949 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10065 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/21896 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/21223 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/html/DOMURL.cpp:
```diff
@@ -39,9 +39,8 @@
-    : m_baseURL(baseURL)
-    , m_url(WTFMove(completeURL))
+inline DOMURL::DOMURL(URL&& completeURL)
+    : m_url(WTFMove(completeURL))
```
ASSERT(m_url.isValid()) could be nice

annevk committed:
Implement URL.parse()
https://bugs.webkit.org/show_bug.cgi?id=271636

Reviewed by Alex Christensen.

This was standardized in https://github.com/whatwg/url/pull/825 and test
coverage was added here:
https://github.com/web-platform-tests/wpt/pull/45248

As a drive-by fix we remove m_baseURL from DOMURL as it does not need it.

* LayoutTests/imported/w3c/web-platform-tests/url/url-statics-parse.any-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/url/url-statics-parse.any.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/url/url-statics-parse.any.js: Added.
(forEach):
(test):
* LayoutTests/imported/w3c/web-platform-tests/url/url-statics-parse.any.worker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/url/url-statics-parse.any.worker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/url/w3c-import.log:
* Source/WebCore/html/DOMURL.cpp:
(WebCore::DOMURL::DOMURL):
(WebCore::DOMURL::create):
(WebCore::parseInternal):
(WebCore::DOMURL::parse):
(WebCore::DOMURL::canParse):
* Source/WebCore/html/DOMURL.h:
* Source/WebCore/html/DOMURL.idl:

Canonical link: https://commits.webkit.org/276656@main
